Sy 1402 make custom schematic symbols feature

emilbon99 opened the pull request:
# Pull Request

Go to the `Preview` tab and select the appropriate template for your pull request:

- [Issue](?expand=1&template=issue.md)
- [RC](?expand=1&template=rc.md)

pjdotson commented on console/src/schematic/symbols/edit/Edit.tsx:
```diff
@@ -0,0 +1,298 @@
+import "@/schematic/symbols/edit/Edit.css";
```
header

pjdotson commented on console/src/schematic/services/palette.tsx:
```diff
@@ -29,4 +30,12 @@ const IMPORT_COMMAND: Palette.Command = {
+  key: "create-symbol",
+  name: "Create Symbol",
+  icon: <CreateIcon />,
+  onSelect: ({ placeLayout }) => placeLayout(createCreateLayout()),
```
what's the point of calling this function `createCreateLayout` when you already have a constant from that file?

emilbon99 replied:
I don't totally understand this. I renamed to `createEditLayout`. I feel like I was following a convention I think you laid out in other parts of the codebase

pjdotson commented on console/src/schematic/symbols/edit/FileDrop.tsx:
```diff
@@ -0,0 +1,93 @@
+import { Flex, Haul, Icon, Status, Text } from "@synnaxlabs/pluto";
```
header

pjdotson commented on console/src/schematic/symbols/edit/HandleList.tsx:
```diff
@@ -0,0 +1,109 @@
+import {
```
header

pjdotson commented on console/src/schematic/symbols/edit/Handles.tsx:
```diff
@@ -0,0 +1,121 @@
+import { type schematic } from "@synnaxlabs/client";
```
header

pjdotson commented on console/src/schematic/symbols/edit/Preview.tsx:
```diff
@@ -0,0 +1,282 @@
+import { type schematic } from "@synnaxlabs/client";
```
header

pjdotson commented on console/src/schematic/symbols/edit/RegionList.tsx:
```diff
@@ -0,0 +1,124 @@
+import { type schematic } from "@synnaxlabs/client";
```
header

pjdotson commented on console/src/schematic/symbols/edit/StateList.tsx:
```diff
@@ -0,0 +1,40 @@
+import { type schematic } from "@synnaxlabs/client";
```
header

pjdotson commented on console/src/schematic/symbols/export.ts:
```diff
@@ -0,0 +1,31 @@
+export interface ExportedSymbol {
+  version: 1;
+  type: "symbol";
+  symbol: schematic.symbol.Symbol;
```
why do you need this type field?

pjdotson commented on console/src/schematic/symbols/export.ts:
```diff
@@ -0,0 +1,31 @@
+    type: "symbol",
+    symbol,
+  };
+  return { data: JSON.stringify(exportData, null, 2), name: symbol.name };
```
I don't think it makes sense to export with the spacing here, we should export it to the smallest size possible.

pjdotson commented on console/src/schematic/symbols/exportGroup.ts:
```diff
@@ -0,0 +1,176 @@
+    return;
+  }
+
+  if (Runtime.ENGINE !== "tauri")
```
you should do this check before making network calls

pjdotson commented on console/src/schematic/symbols/exportGroup.ts:
```diff
@@ -0,0 +1,176 @@
+  if (parentDir == null) return;
+
+  // Create a subdirectory for the symbol group
+  const directoryName = groupName.replace(/[^a-z0-9]/gi, "_");
```
why do the replacing here?

pjdotson commented on console/src/schematic/symbols/exportGroup.ts:
```diff
@@ -0,0 +1,176 @@
+  confirm: Modals.PromptConfirm;
+}
+
+const exportGroup = async ({
```
i feel like some of this logic should be combined with the workspace exporting logic.

pjdotson commented on console/src/schematic/symbols/exportGroup.ts:
```diff
@@ -0,0 +1,176 @@
+  type: "symbol-group";
+  name: string;
+  symbols: SymbolManifest[];
+}
```
can you think some more about duplication in some of the exported data here? we export things to a file directory with a name, but then we also save the group with a name. Each symbol has the file, key, and string in the Symbol manifest, but also a file name of the symbol's file and a name key in the file itself. It seems like there should be less duplication of these things like with the workspace exporting.

pjdotson commented on console/src/schematic/symbols/importGroup.ts:
```diff
@@ -0,0 +1,102 @@
+
+import { Runtime } from "@/runtime";
+
+const manifestZ = z.object({
```
these seems duplicated with the SymbolManifest type in the other file?

pjdotson commented on console/src/schematic/symbols/import.ts:
```diff
@@ -0,0 +1,72 @@
+
+const FILTERS = [{ name: "JSON", extensions: ["json"] }];
+
+export const useImport = (parentGroup?: string): (() => void) => {
```
this seems like sometime that covers a lot of the same code as the other importers

pjdotson reviewed (changes requested):
1. Symbol Search does not work on the toolbar.
2. Creating a symbol group appears to create a symbol group twice.
3. Default sizing when uploading a large file is bad
<img width="1260" height="848" alt="image" src="https://github.com/user-attachments/assets/1fd93723-4b13-4201-86ef-ef419db84820" />

4. Should be able to scroll on mouse wheel with the create symbol modal open
5. Modal triggers tooltip should say "Create" instead of "Save to Synnax"
6. Color selection should clarify that one is the edge color and one is the fill color - I didn't understand this when first using it.
<img width="318" height="94" alt="image" src="https://github.com/user-attachments/assets/bf8b694a-3884-4fb8-85ce-4b00b44d3a21" />
7. name of the symbol should default to the file name uploaded
8. The set the default scale stroke is a little confusing - why wouldn't the default scale always be 100%.
9. A lot of the ux with a larger symbol and uploading it needs to be improved.
10. Should be able to rename the default color group as well.
11. The handles appearance in the editor should size with the scale - when you have a large svg it is impossible to see the handles.
12. Color selection is just not there yet, needs to be a lot more robust + more test cases.
13. When you have a lot of colors + handles the properties piece gets pushed off the modal:
<img width="555" height="804" alt="image" src="https://github.com/user-attachments/assets/7e06a221-f0f0-46f7-aa8c-b3882b402061" />
14. When you hover over a switch we should get a pointer cursor.
15. Console crashed via switching back and forth between the tabs:
<img width="1479" height="906" alt="image" src="https://github.com/user-attachments/assets/464c0918-236b-4be4-944b-a912fd49ab8e" />
16. Default label should be the name not "custom static"
17. It looks like if i have multiple custom groups then the symbol gets added to all of those group?
18. The "no existing ranges, create a range" text is not centered in resources toolbars. This also needs to be added for channels and workspaces.
19. How come the lowest scale I can do is 50%?

emilbon99 commented on pluto/src/index.ts:
```diff
@@ -46,6 +48,8 @@ export { Portal } from "@/portal";
 export { Ranger } from "@/ranger";
 export { Resize } from "@/resize";
+export { Schematic } from "@/schematic";
+export { Symbol as SchematicSymbol } from "@/schematic/symbol";
```
Agreed

emilbon99 commented:
1. Fixed
2. TO-DO
3. Had conversation. Will remain as is for now.
4. TO-DO
5. TO-DO
6. TO-DO
7. TO-DO
8. Had conversation. Will remain as is for now.
10. Fixed
11. See 3
12. Had conversation.
13. TO-Do
14. TO-DO
15. Fixed
16. TO-DO
17. Fixed
18. TO-DO
19. TO-DO

emilbon99 commented on pluto/src/schematic/symbol/external.ts:
```diff
@@ -0,0 +1,13 @@
+export * from "@/schematic/symbol/Custom";
+export * from "@/schematic/symbol/queries";
+export * from "@/schematic/symbol/registry";
+export type * from "@/schematic/symbol/Symbols";
```
I did that so the only thing you can access is the prop types not the symbols themselves.
